src/jit/hwintrinsicxarch.cpp

```diff
@@ -1044,6 +1045,90 @@ GenTree* Compiler::impSSE2Intrinsic(NamedIntrinsic        intrinsic,
             break;
         }
 
+        case NI_SSE2_SetAllVector128:
+        {
+            assert(sig->numArgs == 1);
+            baseType = getBaseTypeOfSIMDType(sig->retTypeSigClass);
+            assert((baseType >= TYP_BYTE) && (baseType <= TYP_DOUBLE));
+            assert(baseType != TYP_FLOAT);
+
+            op1          = impPopStack().val;
+            GenTree* src = nullptr;
+
+            switch (baseType)
+            {
+                case TYP_DOUBLE:
+                {
+                    src     = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, NI_SSE2_SetScalarVector128, baseType, simdSize);
+                    retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, src, gtCloneExpr(src), gtNewIconNode(0),
```

Comment:
You shouldn't be arbitrarily cloning trees.
I think the right thing to use is `Compiler::fgMakeMultiUse`, which will clone it if it is a local, and otherwise create a temp and modify the existing tree to add the assignment. Either way you get back a `GT_LCL_VAR` node to use for the next instance. You can see it in use here: https://github.com/dotnet/coreclr/blob/master/src/jit/morph.cpp#L15008

Reply:
Perhaps, we can change Shuffle intrinsics to support 1-arg form internally to solve this problem.

Reply:
> Perhaps, we can change Shuffle intrinsics to also support 1-arg form internally to solve this problem.

This not necessary since you can use TYP_INT Shuffle as follows to get identical result.

```C++
                    retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, src, gtNewIconNode(0b01000100), NI_SSE2_Shuffle,
                                                       TYP_INT, simdSize);
```

Reply:
@CarolEidt Do you think that usage of `impCloneExpr` for cases `TYP_SHORT` and `TYP_BYTE` as suggested by @AndyAyersMS is OK or I should better switch to `Compiler::fgMakeMultiUse`

Reply:
They are slightly different, and I think `Compiler::fgMakeMultiUse` is the better choice. For one, it has a cleaner interface, because you don't need to pass it a `GenTree**`. The other difference is that `impCloneExpr` will clone any tree that doesn't have side-effects, which is probably not what you want for SIMD, because it will recompute the value, e.g. if it was a SIMD operation.

Reply:
@CarolEidt `impCloneExpr` is used pervasively in the importer for these kinds of decisions. It calls `gtClone` with default last param `false`. So it won't clone anything complex.

You might be thinking of the behavior of `gtCloneExpr` which does much more more extensive cloning.

Reply:
@CarolEidt @AndyAyersMS 

It appears that in case HW intrinsics `Compiler::fgMakeMultiUse` works much better. I have got significantly better codegen except for unnecessary use of XMM1 register:

Word sized operands:
```assembly
       480FBECE             movsx    rcx, sil
       C4E1796EC1           vmovd    xmm0, xrcx
       C4E17960C8           vpunpcklbw xmm1, xmm0, xmm0
       C4E17161C1           vpunpcklwd xmm0, xmm1, xmm1
       C4E17970C000         vpshufd  xmm0, xmm0, 0

// SSE legacy encoding

       480FBFCE             movsx    rcx, si
       660F6EC1             movd     xmm0, xrcx
       0F28C8               movaps   xmm1, xmm0
       660F61C8             punpcklwd xmm1, xmm0
       660F70C100           pshufd   xmm0, xmm1, 0
```

Byte sized operands:
```assembly
       400FB6CE             movzx    rcx, sil
       C4E1796EC1           vmovd    xmm0, xrcx
       C4E17960C8           vpunpcklbw xmm1, xmm0, xmm0
       C4E17161C1           vpunpcklwd xmm0, xmm1, xmm1
       C4E17970C000         vpshufd  xmm0, xmm0, 0

// SSElegacy encoding

       480FBECE             movsx    rcx, sil
       660F6EC1             movd     xmm0, xrcx
       0F28C8               movaps   xmm1, xmm0
       660F60C8             punpcklbw xmm1, xmm0
       0F28C1               movaps   xmm0, xmm1
       660F61C1             punpcklwd xmm0, xmm1
       660F70C000           pshufd   xmm0, xmm0, 0
```

Reply:
Do you have any hints how to block second XMM register use and get rid of `movaps   xmm1, xmm0` operations?

Reply:
If it's really better to go via `fgMakeMultiUse` it would be good to understand why; like I said the rest of the importer uses `impCloneExpr` pretty heavily.

Reply:
> impCloneExpr is used pervasively in the importer for these kinds of decisions. It calls gtClone with default last param false. So it won't clone anything complex.

@AndyAyersMS - `impCloneExpr` actually passes `true` to `gtClone`, but I see that it will only clone simple trees. So that may be the way to go.

Reply:
> Do you have any hints how to block second XMM register use and get rid of movaps xmm1, xmm0 operations?

I would have to look at a jitdump to see what's going on.

Reply:
@CarolEidt 

These are dumps comprising only `short`, `ushort`, `byte` and `sbyte` tests for AVX and SSE codegen:

[SetAllVector128Ro.txt](https://github.com/dotnet/coreclr/files/1782987/SetAllVector128Ro.txt)
[SetAllVector128RoNoAVXDump.txt](https://github.com/dotnet/coreclr/files/1782988/SetAllVector128RoNoAVXDump.txt)

Reply:
Ooops, yeah, it passes true, so it will clone add/sub/addr etc.

If there is something deficient about it it would be good to know what it is since it is heavily used. On the surface I don't see anything obvious.